cpp/src/arrow/adapters/orc/jni/jni_wrapper.cpp

```diff
@@ -0,0 +1,243 @@
+  if (ioExceptionClass != nullptr) {
+    if (env->ThrowNew(ioExceptionClass, e.what())) {
+      // Failed to new IOException. This means another error has occurred in Java
+      // We just propagate this error to caller by doing nothing.
+      ARROW_LOG(ERROR) << "Error occurred when throwing IOException";
+    }
+  } else {
+    ARROW_LOG(ERROR) << "Error occurred when getting IOException class";
+  }
+}
+
+void ThrowJavaException(JNIEnv* env, const std::string& message) {
+  jclass exception = env->FindClass("java/lang/Exception");
+  if (exception != nullptr) {
+    env->ThrowNew(exception, message.c_str());
+  } else {
+    throw std::runtime_error("Can't find java/lang/Exception class");
+  }
+}
+
+jfieldID GetFieldId(JNIEnv* env, jclass this_class, const std::string& sig) {
+  jfieldID ret = env->GetFieldID(this_class, sig.c_str(), "J");
+  if (ret == nullptr) {
+    ThrowJavaException(env, "Unable to get java class field: " + sig);
+  }
+
+  return ret;
+}
+
+std::unique_ptr<ORCFileReader>* GetNativeReader(JNIEnv* env, jobject this_obj) {
+  jlong reader = env->GetLongField(
+      this_obj, GetFieldId(env, env->GetObjectClass(this_obj), "nativeReaderAddress"));
+  return reinterpret_cast<std::unique_ptr<ORCFileReader>*>(reader);
+}
+
+std::shared_ptr<RecordBatchReader>* GetStripeReader(JNIEnv* env, jobject this_obj) {
+  jlong reader = env->GetLongField(
+      this_obj,
+      GetFieldId(env, env->GetObjectClass(this_obj), "nativeStripeReaderAddress"));
+  return reinterpret_cast<std::shared_ptr<RecordBatchReader>*>(reader);
+}
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+JNIEXPORT jboolean JNICALL Java_org_apache_arrow_adapter_orc_OrcReaderJniWrapper_open(
+    JNIEnv* env, jobject this_obj, jstring file_path) {
+  std::shared_ptr<arrow::io::ReadableFile> in_file;
+  const char* str = env->GetStringUTFChars(file_path, nullptr);
+  std::string path(str);
+  env->ReleaseStringUTFChars(file_path, str);
+
+  arrow::Status ret;
+  if (path.find("hdfs://") == 0) {
+    return false;
+  } else {
+    ret = arrow::io::ReadableFile::Open(path, &in_file);
+  }
+
+  if (ret.ok()) {
+    auto reader = new std::unique_ptr<ORCFileReader>();
+
+    ret = ORCFileReader::Open(
+        std::static_pointer_cast<arrow::io::RandomAccessFile>(in_file),
+        arrow::default_memory_pool(), reader);
+
+    if (!ret.ok()) {
+      delete reader;
+      ThrowJavaIOException(env, std::invalid_argument("Failed open file" + path));
+    }
+
+    env->SetLongField(
+        this_obj, GetFieldId(env, env->GetObjectClass(this_obj), "nativeReaderAddress"),
+        reinterpret_cast<long>(reader));
+  }
+
+  return ret.ok();
+}
+
+JNIEXPORT void JNICALL Java_org_apache_arrow_adapter_orc_OrcReaderJniWrapper_close(
+    JNIEnv* env, jobject this_obj) {
+  delete GetNativeReader(env, this_obj);
+}
+
+JNIEXPORT jboolean JNICALL Java_org_apache_arrow_adapter_orc_OrcReaderJniWrapper_seek(
+    JNIEnv* env, jobject this_obj, jint row_number) {
```

Comment:
not too familiar with JNI but should this be jlong?

Reply:
here it is jint because we are passing in an int parameter